script/core/definition.lua

```diff
@@ -7,18 +7,70 @@ local rpath      = require 'workspace.require-path'
 local jumpSource = require 'core.jump-source'
 local wssymbol   = require 'core.workspace-symbol'
 
-local function sortResults(results)
+--- @param s string
+--- @return string[]
+local function split(s)
+    local r = {}
+    s:gsub('[^/]+', function (w)
+        r[#r+1] = w:gsub("~1", "/"):gsub("~0", "~")
+    end)
+    return r
+end
+
+--- Returns the Levenshtein distance between the two given string arrays
+--- @param a string[]
+--- @param b string[]
+--- @return number
+local function levenshtein_distance(a, b)
+  local a_len, b_len = #a, #b
+  local matrix = {} --- @type integer[][]
+
+  -- Initialize the matrix
+  for i = 1, a_len + 1 do
+      matrix[i] = { [1] = i }
+  end
+
+  for j = 1, b_len + 1 do
+      matrix[1][j] = j
+  end
+
+  -- Compute the Levenshtein distance
+  for i = 1, a_len do
+      for j = 1, b_len do
+          local cost = (a[i] == b[j]) and 0 or 1
+          matrix[i + 1][j + 1] =
+              math.min(matrix[i][j + 1] + 1, matrix[i + 1][j] + 1, matrix[i][j] + cost)
+       end
+  end
+
+  -- Return the Levenshtein distance
+  return matrix[a_len + 1][b_len + 1]
+end
```

Comment:
I see many for loops here, and the `path_similarity_ratio()` is even called inside a `table.sort` comparator function. Would this incur a significant performance impact? 😳

---

Seems better to cache the result of each `path_similarity_ratio(uri, u)` ?
- `uri` is constant in this context, only `u` is a variable here
- maybe something like this
```lua
local ratioCache = {}
table.sort(results, function (a, b)
    local u1 = guide.getUri(a.target)
    local u2 = guide.getUri(b.target)
    if u1 == u2 then
        return a.target.start < b.target.start
    else
        ratioCache[u1] = ratioCache[u1] or path_similarity_ratio(uri, u1)
        ratioCache[u2] = ratioCache[u2] or path_similarity_ratio(uri, u2)
        return ratioCache[u1] < ratioCache[u2]
    end
end)
```

Reply:
95% of the time `results <= 2` which means the cache never hits, but I've added it anyway for good measure.